kernel/src/checkpoint/mod.rs

```diff
@@ -2,7 +2,462 @@
+/// checkpoint by writing the `_last_checkpoint` file.
+pub struct CheckpointWriter {
+    /// Reference to the snapshot of the table being checkpointed
+    pub(crate) snapshot: Arc<Snapshot>,
+    /// Note: `Arc<AtomicI64>` provides shared mutability for our counters, allowing the
+    /// returned actions iterator from `.checkpoint_data()` to update the counters,
+    /// and the [`CheckpointWriter`] to read them during `.finalize()`
+    /// Counter for total actions included in the checkpoint
+    pub(crate) total_actions_counter: Arc<AtomicI64>,
+    /// Counter for Add actions included in the checkpoint
+    pub(crate) add_actions_counter: Arc<AtomicI64>,
+}
+
+impl CheckpointWriter {
+    /// Creates a new CheckpointWriter with the provided checkpoint data and counters
+    pub(crate) fn new(snapshot: Arc<Snapshot>) -> Self {
+        Self {
+            snapshot,
+            total_actions_counter: Arc::new(AtomicI64::new(0)),
+            add_actions_counter: Arc::new(AtomicI64::new(0)),
+        }
+    }
+
+    /// Retrieves the checkpoint data and path information
+    ///
+    /// This method is the core of the checkpoint generation process. It:
+    /// 1. Determines whether to write a V1 or V2 checkpoint based on the table's
+    ///    `v2Checkpoints` feature support
+    /// 2. Reads actions from the log segment using the checkpoint read schema
+    /// 3. Filters and deduplicates actions for the checkpoint
+    /// 4. Chains the checkpoint metadata action if writing a V2 spec checkpoint
+    ///    (i.e., if `v2Checkpoints` feature is supported by table)
+    /// 5. Generates the appropriate checkpoint path
+    ///
+    /// # Important: The returned data should be written to persistent storage by the
+    /// caller before calling `finalize()` otherwise data loss may occur.
+    ///
+    /// # Returns: [`CheckpointData`] containing the checkpoint path and data to write
+    pub fn checkpoint_data(&mut self, engine: &dyn Engine) -> DeltaResult<CheckpointData> {
+        let is_v2_checkpoints_supported = self
+            .snapshot
+            .table_configuration()
+            .is_v2_checkpoint_supported();
+
+        let actions = self.snapshot.log_segment().read_actions(
+            engine,
+            CHECKPOINT_ACTIONS_SCHEMA.clone(),
+            CHECKPOINT_ACTIONS_SCHEMA.clone(),
+            None,
+        )?;
+
+        // Create iterator over actions for checkpoint data
+        let checkpoint_data = CheckpointLogReplayProcessor::new(
+            self.total_actions_counter.clone(),
+            self.add_actions_counter.clone(),
+            self.deleted_file_retention_timestamp()?,
+        )
+        .process_actions_iter(actions);
+
+        let version = self.snapshot.version().try_into().map_err(|e| {
+            Error::checkpoint_writer(format!(
+                "Failed to convert checkpoint version from u64 {} to i64: {}",
+                self.snapshot.version(),
+                e
+            ))
```

Comment:
Nit:
```suggestion
            Error::CheckpointWriter(format!(
                "Failed to convert checkpoint version from u64 {} to i64: {}",
                self.snapshot.version(),
                e
            ))
```

The `checkpoint_writer` type functions are when you want to pass something that's not a `String` and they call `to_string()`. By doing the format here _and_ using that function you're doing unnecessary work (although `to_string` is probably basically a no-op on a `String`).

Reply:
Got it, thanks! That makes sense, will switch to using Error::CheckpointWriter directly to avoid the extra formatting work.